### ISSUE TITLE
Custom DSC resource not applying because a parameter is not found

### ISSUE THREAD

**Issue by bsmienk** [puppetlabs/puppetlabs-dsc#286]
Hello,

A few days ago I created an issue for a problem I had while trying to create a custom DSC module for WDS. I was very happy with the help I was offered here, unfortunately I have found another issue. I am unsure if it is an actual bug or if I am doing something wrong myself. It seems likely that I am doing something wrong myself, however I just can't find the issue.

At the moment I have created quite a few resources and have been testing/experimenting with them.
There is one custom DSC resource that is causing me some problems. At first the resource was working as intended. However, as I wanted to add a feature to the resource I had to add a new parameter. Which, I think, I did correctly. However whenever I try to use this new parameter in my puppet profile I am getting an error at the machine that is trying to apply it. The files can be found in this repo;
https://github.com/bsmienk/custom_dsc_modules/tree/master/cWDS/DSCResources/cWDS_AddDriverGroup

The error is as follows:
```
Info: Loading facts
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: no parameter named 'dsc_setapplicability' at
/etc/puppetlabs/code/environments/puppetdev/modules/profile/manifests/wds/server.pp:81 on Dsc_cwds_adddrivergroup[Ad
dRHEVGroup] at /etc/puppetlabs/code/environments/puppetdev/modules/profile/manifests/wds/server.pp:81 on node <nodename>
```

Things I tried:
- Rebuilding the resources on the puppet master
- Removing the ProgramData\PuppetLabs\puppet\cache directory on the node
- Changing the parameter name (error changed to display the new parameter name too)
- Redownloading the files from github and then rebuilding the files

If I don't  use the parameter it all works as intended. But using this parameter does not seem to be allowed.

I have described the parameter in the schema file like this:
```
[Write, ValueMap{"Matched", "All"},Values{"Matched", "All"}] string SetApplicability;
```
I am using the follow code in my profile on the puppet master
```
dsc_cwds_adddrivergroup{'AddRHEVGroup':
  dsc_drivergroup => 'RHEV',
  dsc_enabled => 'Yes',
  dsc_setapplicability => 'Matched',
  dsc_ensure => 'Present'
}
```
I think I have found the related type file on the node in this directory:
```
C:\ProgramData\PuppetLabs\puppet\cache\lib\puppet\type\dsc_cwds_adddrivergroup.rb
```
In that File I have found the following code:
```
  # Name:         SetApplicability
  # Type:         string
  # IsMandatory:  False
  # Values:       ["Matched", "All"]
  newparam(:dsc_setapplicability) do
    def mof_type; 'string' end
    def mof_is_embedded?; false end
    desc "SetApplicability - Valid values are Matched, All."
    validate do |value|
      unless value.kind_of?(String)
        fail("Invalid value '#{value}'. Should be a string")
      end
      unless ['Matched', 'matched', 'All', 'all'].include?(value)
        fail("Invalid value '#{value}'. Valid values are Matched, All")
      end
    end
  end
```

I hope someone can explain why this parameter is not working. Am I doing something wrong? Or is there another problem?

P.s. When I am building all the types using 'bundle exec rake dsc:types:build' I get a warning for every resource, saying that my schema.mof files are expected ot have CRLF line endings (which I believe they do). Despite this, everything else works.

P.s.2. I have just tried using the DSC resource locally on the node and this works fine.
```
Configuration Test {
    
    Import-DscResource -ModuleName 'cWDS'

    Node localhost{

        cWDS_AddDriverGroup 'Test' {
            DriverGroup = 'Test'
            SetApplicability = 'Matched'
            Ensure = 'Present'
            Enabled = 'Yes'
        }
    }
}

Test -OutputPath 'C:\Users\Administrator.CORE\Desktop'
Start-DscConfiguration -Path 'C:\Users\Administrator.CORE\Desktop' -Wait -Verbose -Force
```
Results in:
```
VERBOSE: Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = SendConfigurationApply,'className' = MSFT_DSCLocalConfigurationManager,'namespace
Name' = root/Microsoft/Windows/DesiredStateConfiguration'.
VERBOSE: An LCM method call arrived from computer CORE-O-WDS01 with user sid S-1-5-21-2749254005-1462001878-1174754042-500.
VERBOSE: [CORE-O-WDS01]: LCM:  [ Start  Set      ]
VERBOSE: [CORE-O-WDS01]: LCM:  [ Start  Resource ]  [[cWDS_AddDriverGroup]Test]
VERBOSE: [CORE-O-WDS01]: LCM:  [ Start  Test     ]  [[cWDS_AddDriverGroup]Test]
VERBOSE: [CORE-O-WDS01]: LCM:  [ End    Test     ]  [[cWDS_AddDriverGroup]Test]  in 0.3590 seconds.
VERBOSE: [CORE-O-WDS01]: LCM:  [ Skip   Set      ]  [[cWDS_AddDriverGroup]Test]
VERBOSE: [CORE-O-WDS01]: LCM:  [ End    Resource ]  [[cWDS_AddDriverGroup]Test]
VERBOSE: [CORE-O-WDS01]: LCM:  [ End    Set      ]
VERBOSE: [CORE-O-WDS01]: LCM:  [ End    Set      ]    in  0.6250 seconds.
VERBOSE: Operation 'Invoke CimMethod' complete.
VERBOSE: Time taken for configuration job to complete is 0.643 seconds
```

**Comment by jpogran** [comment]
Thanks again for a more than complete bug report and repo steps, it really helps.

When you run DSC using `Start-DscConfiguration`, it resolves the DSC Resources using the `$env:PSModulePath` variable. We puppet executes DSC using `Invoke-DscResource`, it provides the full path to the DSC Resource which is not in `$env:PSModulePath` but in the same folder the puppet dsc module is located in. We do this to isolate the DSC Resources we bundle into the dsc module from conflicting or overwriting any a user has deployed to their system.

So, in summary there are two parts that need to be copied down: the puppet type file (`dsc_cwds_adddrivergroup.rb`) and the actual DSC Resource module (`cWDS_AddDriverGroup.psm1`). They will be located in the location where your puppet modules are synced to when the master pushed down to the target node.

In my example in the previous ticket, we can see mine were located in `C:/puppetlabs/modules/dsc/lib/puppet_x/dsc_resources/xCMC_WDS/xCMC_WDS.psd1`

```
$invokeParams = @{
  Name          = 'Configure_WDS'
  Method        = 'test'
  Property      = @{
    configurationname = 'StandardConfiguration'
    confighash = [CimInstance](New-CimInstance -ClassName 'MSFT_KeyValuePair' -ClientOnly -Property @{'Answer clients: Yes' = '/AnswerClients:All'})
  }
  ModuleName = @{
    ModuleName      = "C:/puppetlabs/modules/dsc/lib/puppet_x/dsc_resources/xCMC_WDS/xCMC_WDS.psd1"
    RequiredVersion = "1.0"
  }
}
```

If you run puppet on the target node using debug, you will see the path your module is in. Make sure that location is up to date and has the new files. I'm relatively certain this is a file sync issue from puppet master to target node, and not anything to do with the module itself as you demonstrated above it built successfully. You may have to increment the metatadata.json version to indicate to the puppet master to copy out new files, I will check with some colleagues to double check.

**Comment by bsmienk** [comment]
I have tried fixing this and I don't have a 'modules\dsc' directory anywhere. I have found multiple modules folders but they are all empty:
- 'C:\Program Files\Puppet Labs\Puppet\puppet\modules'
- 'C:\ProgramData\PuppetLabs\code\modules'
- 'C:\ProgramData\PuppetLabs\code\environments\production\modules'
They are all empty. It is also strange, in my opinion, that there is no 'C:\ProgramData\PuppetLabs\code\environments\puppetdev' directory, since that is the environment it is assigned to. I would assume the modules to be saved in that directory but it doesn't exist.

On the other hand, I have found the .psm1 file in another directory:
- 'C:\ProgramData\PuppetLabs\puppet\cache\lib\puppet_x\dsc_resources\cWDS\DSCResources\cWDS_AddDriverGroup'
In this directory I have found the correct psm1 and mof file, including the parameter.

Also, I tried deleting the hard disk of the node and reinstalling windows + puppet, and the problem persists.
